### PR TITLE
(7/n) Support 2D Parallelism - TP Fabric Docs

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds a basic tutorial for tensor parallelism in Lightning. Since tensor parallelism is a relatively new concept in PyTorch and requires some understanding of your model, I'm adding a conceptual guide with illustrations to give enough context to the code examples. 

Once this is reviewed, I will copy-paste this for the Trainer docs (separate PR). 

----
📚 Documentation preview 📚: https://pytorch-lightning--19884.org.readthedocs.build/en/19884/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: docs/source-fabric/advanced/model_parallel/index.rst [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+        </div>
+    </div>
+
+
+----
+
+
+*********************
+Parallelisms compared
```
Comment (awaelchli) [inline docs/source-fabric/advanced/model_parallel/index.rst:99]: I initially wrote this first, but now I'm not so sure anymore if it's the best way to present it. 
What do you like more? The text intro at the top or the explicit pros/cons bullet points here?

File: docs/source-fabric/advanced/model_parallel/tp.rst [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+        fabric.print(f"Peak memory usage: {torch.cuda.max_memory_reserved() / 1e9:.02f} GB")
+
+|
+
+When measuring the peak memory consumption, we should see that doubling the number of GPUs reduces the memory consuption roughly by half:
+
+
+.. list-table::
+   :widths: 20 20 20 20 20
```
Comment (awaelchli) [inline docs/source-fabric/advanced/model_parallel/tp.rst:206]: I want to do better benchmarks in the future, comparing different configurations. See this as a V0.

File: docs/source-fabric/advanced/model_parallel/fsdp.rst [hunk 3]
```diff
@@ -1,12 +1,9 @@
...
+Training models with billions of parameters with FSDP
+#####################################################
 
 Use Fully Sharded Data Parallel (FSDP) to train large models with billions of parameters efficiently on multiple GPUs and across multiple machines.
 
-.. note:: This is an experimental feature.
-
-
 Today, large models with billions of parameters are trained with many GPUs across several machines in parallel.
 Even a single H100 GPU with 80 GB of VRAM (the biggest today) is not enough to train just a 30B parameter model (even with batch size 1 and 16-bit precision).
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/fsdp.rst:8]: Since there's the GB200 now:

```suggestion
Even a single H100 GPU with 80 GB of VRAM (one of the biggest today) is not enough to train just a 30B parameter model (even with batch size 1 and 16-bit precision).
```

File: docs/source-fabric/advanced/model_parallel/index.rst [hunk 4]
```diff
@@ -0,0 +1,162 @@
+###########################################
+Training models with billions of parameters
+###########################################
+
+Today, large models with billions of parameters are trained with many GPUs across several machines in parallel.
+Even a single H100 GPU with 80 GB of VRAM (the biggest today) is not enough to train just a 30B parameter model (even with batch size 1 and 16-bit precision).
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/index.rst:6]:
```suggestion
Even a single H100 GPU with 80 GB of VRAM (one of the biggest today) is not enough to train just a 30B parameter model (even with batch size 1 and 16-bit precision).
```

File: docs/source-fabric/advanced/model_parallel/index.rst [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+The memory consumption for training is generally made up of
+
+1. the model parameters,
+2. the layer activations (forward),
+3. the gradients (backward) and
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/index.rst:11]:
```suggestion
3. the gradients (backward),
```

File: docs/source-fabric/advanced/model_parallel/index.rst [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+4. the optimizer states (e.g., Adam has two additional exponential averages per parameter).
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/index.rst:12]:
```suggestion
4. the optimizer states (e.g., Adam has two additional exponential averages per parameter) and
5. model outputs and loss for each batch.
```

The loss is kind of related to the layer activations, but I was thinking of it explicitly here due to the cross entropy creating spikes in LitGPT, which is why we added the chunked cross entropy.

File: docs/source-fabric/advanced/model_parallel/index.rst (hunk 1 above)
Comment (rasbt) [reply docs/source-fabric/advanced/model_parallel/index.rst:99]: I think the text reads very well, I would not remove it. However, optionally a table or section with short pro/con bullet points can be nice to summarize the text. But the text is nice because it also explains more. Maybe you could add the bullet points below each text section and make them shorter.

File: docs/source-fabric/advanced/model_parallel/tp.rst [hunk 7]
```diff
@@ -0,0 +1,280 @@
+##################
+Tensor Parallelism
+##################
+
+Tensor parallelism is a technique for training large models by distributing layers across multiple devices, improving memory management and efficiency by reducing inter-device communication.
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/tp.rst:5]: ~~In addition to distributing layers, is it also distributing the matrix multiplication etc (or does it move the matrices to the same device when multiplying?~~ (NVM, you answer it below)

If it does the matrix multiplication in a distributed way, then a key point to mention here could be that this is useful when there are large layers (as opposed to a large model with lots of layers but small layers)

File: docs/source-fabric/advanced/model_parallel/tp.rst [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+***********************************
+Apply tensor parallelism to a model
+***********************************
+
+To apply tensor parallelism to a model with Fabric, you need a good understanding of your model's architecture to make the decision of where to apply the parallel styles you've seen above.
+Let's start with a simple MLP toy example:
+
+.. code-block:: python
+
+    import torch
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/tp.rst:74]:
```suggestion
```
You are not using the `torch` import below so you could technically remove it

File: docs/source-fabric/advanced/model_parallel/tp.rst [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        strategy = ModelParallelStrategy(parallelize_fn=parallelize_feedforward)
+        fabric = L.Fabric(accelerator="cuda", devices=2, strategy=strategy)
+        fabric.launch()
+
+        # Initialize the model
+        model = FeedForward(8192, 8192)
+        model = fabric.setup(model)
+
+        # Define the optimizer
+        optimizer = torch.optim.AdamW(model.parameters(), lr=3e-3, foreach=True)
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/tp.rst:181]: Regarding 

> foreach=True

every time I read your tutorials, I learn something new 😊

File: docs/source-fabric/advanced/model_parallel/tp.rst [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+        # Simplified training loop
+        for i, batch in enumerate(dataloader):
+            output = model(batch)
+            loss = output.sum()
+            fabric.backward(loss)
+            optimizer.step()
+            optimizer.zero_grad()
+            fabric.print(f"Iteration {i} complete")
+
+        fabric.print(f"Peak memory usage: {torch.cuda.max_memory_reserved() / 1e9:.02f} GB")
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/tp.rst:198]: Why do we not use `torch.cuda.max_memory_allocated()`? Just a naive question how it's different from `torch.cuda.max_memory_reserved()`. Isn't `torch.cuda.max_memory_reserved()` showing a number that is usually larger if you have more space? E.g., it may allocate more memory on a larger GPU than a smaller GPU even though the code could run on both GPUs?

File: docs/source-fabric/advanced/model_parallel/tp.rst [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+
+----
+
+
+***************************
+Data-loading considerations
+***************************
+
+In a tensor-parallelized model, it is important that the model receives an identical input on each GPU.
```
Comment (rasbt) [inline docs/source-fabric/advanced/model_parallel/tp.rst:230]: Intuitively, I would then think it can't be combined with data parallelism. But below you mention it can be combined, which is cool. Maybe you can say a few words here to clarify that it can still be used with data parallelism, and in short how it is handled (if it can be said in short).

**Review by rasbt — COMMENTED** [review]
The docs read very smoothly and are very clear. Actually, this is a great documentation for people who want to learn about the different multi-GPU training paradigms. I don't think there are any other resources on the internet that talk about and compare them all. Great work!

File: docs/source-fabric/advanced/model_parallel/tp.rst (hunk 10 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/tp.rst]: You are right yes, it should be allocated.

File: docs/source-fabric/advanced/model_parallel/tp.rst (hunk 9 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/tp.rst:180]: To be fair, this is copy-paste and not directly related to TP, so maybe it should actually be removed :))

File: docs/source-fabric/advanced/model_parallel/tp.rst (hunk 11 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/tp.rst:229]: I personally think this is ok. At this point, the tutorial is focused on TP alone and I wanted to emphasize on this detail. In the future, I also want to write a section here about inference, where of course data-parallel is out of question anyway.
To explain how TP + DP is combined, I would rather rely on the separate page where we link (submitting a PR shortly) to fully explain it in detail.

File: docs/source-fabric/advanced/model_parallel/index.rst (hunk 1 above)
Comment (lantiga) [reply docs/source-fabric/advanced/model_parallel/index.rst:100]: I love the section personally, it's clear and factual.

File: docs/source-fabric/advanced/model_parallel/tp.rst (hunk 9 above)
Comment (lantiga) [reply docs/source-fabric/advanced/model_parallel/tp.rst:180]: yes maybe this would lead readers to think that `foreach=True` is somehow related

File: docs/source-fabric/advanced/model_parallel/tp.rst [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+     - 2 GPUs
+     - 4 GPUs
+     - 8 GPUs
+   * - Memory per GPU
+     - 4.04 GB
+     - 2.03 GB
+     - 1.02 GB
+     - 0.60 GB
+
+Beyond this toy example, we recommend you study our `LLM Tensor Parallel Example (Llama 2) <https://github.com/Lightning-AI/pytorch-lightning/tree/master/examples/fabric/tensor_parallel>`_.
```
Comment (lantiga) [inline docs/source-fabric/advanced/model_parallel/tp.rst:219]: This is both Llama 2 and Llama 3 right from TorchTitan right? If they are the same we could just say Llama 3 to make it more up to date.

**Review by lantiga — APPROVED** [review]
Digged the new doc section @awaelchli

File: docs/source-fabric/advanced/model_parallel/tp.rst (hunk 9 above)
Comment (rasbt) [reply docs/source-fabric/advanced/model_parallel/tp.rst:180]: Yeah, I think it makes sense for this architecture in general, because it's not memory bound, but introducing it here could somehow make readers believe it is required or so. So maybe just dropping it would be better.

**Review by rasbt — APPROVED** [review]

File: docs/source-fabric/advanced/model_parallel/tp.rst (hunk 12 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/tp.rst]: That's right, they are architecturally the same so we can name this Llama 3 👍

File: docs/source-fabric/advanced/model_parallel/tp.rst (hunk 9 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/tp.rst]: removed 👍

**Review by carmocca — APPROVED** [review]

File: docs/source-fabric/advanced/model_parallel/index.rst [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+**Fully-Sharded Data Parallel (FSDP)**
+
+.. raw:: html
+
+    <ul class="no-bullets">
+        <li>✅ &nbsp; No model code changes required </li>
+        <li>✅ &nbsp; Training with very large batch sizes (batch size scales with number of GPUs)</li>
+        <li>✅ &nbsp; Model (weights, optimizer state, activations) gets distributed across all GPUs </li>
+        <li>❗ &nbsp; Forward/backward computation requires a single layer must fit into a GPU </li>
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/index.rst:124]: Right
These bullet points are definitely not super detailed, they are meant to give a big picture overview. With the single layer, I want to express what the extremes are. In this context, I think if there are non-fsdp wrapped layers then they are small enough to not matter for this argument.

[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source-fabric/advanced/model_parallel/index.rst:124]: Embeddings can a very large percentage for models in the low billions with large vocabulary sizes (like gemma)

[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/index.rst:124]: Is this better?

> A single FSDP layer when gathered during forward/backward must fit into the GPU

[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source-fabric/advanced/model_parallel/index.rst:124]: It is better

File: docs/source-fabric/advanced/model_parallel/index.rst [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+**************************
+What is Model Parallelism?
+**************************
+
+There are different types of model parallelism, each with its own trade-offs.
+
+**Fully Sharded Data Parallelism (FSDP)** shards both model parameters and optimizer states across multiple GPUs, significantly reducing memory usage per GPU.
+This method, while highly memory-efficient, involves frequent synchronization between GPUs, introducing communication overhead and complexity in implementation.
+FSDP is advantageous when memory constraints are the primary issue, provided there are high-bandwidth interconnects to minimize latency.
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/index.rst:32]: I will try to make the lines less blurred here by changing the intro in a future PR.
